src/parser/common.rs

```diff
@@ -155,6 +155,43 @@ pub fn parse_gate_modifier<'a>(input: ParserInput<'a>) -> ParserResult<'a, GateM
     ))
 }
 
+/// Parse matrix used to define gate with `DEFGATE`.
+pub fn parse_matrix<'a>(input: ParserInput<'a>) -> ParserResult<'a, Vec<Vec<Expression>>> {
+    let (input, matrix) = separated_list0(
+        token!(NewLine),
+        preceded(
+            many1(token!(Indentation)),
```

Comment:
I believe this should be exactly one indentation. Likewise with permutation below. See `parse_block_instruction` and just swap `parse_instruction` for the expected line contents here.

On that note, `parse_indented_block` could be its own useful combinator here which takes the line's parser as an argument... would be used in 4 places as of this PR:

* permutation
* matrix
* waveform
* block_instruction

Reply:
Yeah, this would be a good one! Time might be hard to come by in the near future, but let's earmark this one for a later 1x1. We can punt it for this PR.